core/src/main/java/dk/alexandra/fresco/lib/compare/bool/eq/BinaryEqualityProtocolImpl.java

```diff
@@ -3,119 +3,85 @@
  *
  * @author Kasper Damgaard
  */
-public class BinaryEqualityProtocolImpl implements BinaryEqualityProtocol {
+public class BinaryEqualityProtocolImpl implements ComputationBuilderBinary<SBool> {
 
-  private BasicLogicFactory factory;
-  private SBool[] inLeft;
-  private SBool[] inRight;
-  private SBool out;
+  private List<Computation<SBool>> inLeft;
+  private List<Computation<SBool>> inRight;
   private final int length;
 
-  private SBool[] xnorOuts;
-
-  private int step = 1;
-  private boolean xnorDone = false;
-
-  private ProtocolProducer curPP = null;
-
-  public BinaryEqualityProtocolImpl(SBool[] inLeft, SBool[] inRight,
-      SBool out, BasicLogicFactory factory) {
-    this.factory = factory;
+  public BinaryEqualityProtocolImpl(List<Computation<SBool>> inLeft,
+      List<Computation<SBool>> inRight) {
     this.inLeft = inLeft;
     this.inRight = inRight;
-    this.out = out;
-    if (inLeft.length != inRight.length) {
-      throw new IllegalArgumentException(
-          "Binary strings must be of equal length");
+    if (inLeft.size() != inRight.size()) {
+      throw new IllegalArgumentException("Binary strings must be of equal length");
     }
-    this.length = inLeft.length;
+    this.length = inLeft.size();
   }
 
   @Override
-  public void getNextProtocols(ProtocolCollection protocolCollection) {
-    if (xnorOuts == null) {
-      xnorOuts = new SBool[length];
-      for (int i = 1; i < length; i++) {
-        xnorOuts[i] = factory.getSBool();
+  public Computation<SBool> build(SequentialBinaryBuilder builder) {
+    return builder.par(par -> {
+      BinaryBuilder bb = par.binary();
+      List<Computation<SBool>> xors = new ArrayList<>();
+      for (int i = 0; i < length; i++) {
+        xors.add(bb.xor(inLeft.get(i), inRight.get(i)));
       }
-      xnorOuts[0] = out;
-    }
-    if (curPP == null) {
-      if (!xnorDone) {
-        ParallelProtocolProducer parXOR = new ParallelProtocolProducer();
-        ParallelProtocolProducer parNOT = new ParallelProtocolProducer();
-        for (int i = 0; i < length; i++) {
-          parXOR.append(factory.getXorProtocol(inLeft[i], inRight[i],
-              xnorOuts[i]));
-          parNOT.append(factory.getNotProtocol(xnorOuts[i],
-              xnorOuts[i]));
-        }
-        curPP = new SequentialProtocolProducer(parXOR, parNOT);
-      } else {
-        ParallelProtocolProducer parAND = new ParallelProtocolProducer();
-        int i = 0;
-        while (i + step < length) {
-          ProtocolProducer and = factory.getAndProtocol(xnorOuts[i],
-              xnorOuts[i + step], xnorOuts[i]);
-          parAND.append(and);
-          i += 2 * step;
-        }
-        curPP = parAND;
+      return () -> xors;
+    }).par((xors, par) -> {
+      List<Computation<SBool>> xnors = new ArrayList<>();
+      for (int i = 0; i < length; i++) {
+        xnors.add(par.binary().not(xors.get(i)));
```

Comment:
This seems to be equivalent to doing xnor in the par.advanced()

Either remove that method or use it here?